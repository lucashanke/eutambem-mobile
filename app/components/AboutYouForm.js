import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { View, Text } from 'react-native';

import appStyles, { RED } from '../styles';

import { Button, CheckBox, Picker, TextInput, SectionHeader, SectionText } from './common';
import { formWrapper, loading } from './hoc';

export class AboutYouForm extends Component {
  constructor(props) {
    super(props);
    this.state = {
      formData: {
        gender: { value: '', valid: false },
        skinColor: { value: '', valid: false },
        age: { value: '', valid: true },
        wage: { value: '', valid: true },
        email: { value: '', valid: false },
        name: { value: '', valid: true },
        sexualOrientation: { value: '', valid: true },
      },
      acceptedPolicies: false,
      containErrors: false,
    };
  }

  updateFormDataValue = (key, value, valid) => {
    const { formData } = this.state;
    if (key in formData) {
      formData[key] = { value, valid };
      this.setState({ formData });
    } else {
      console.error('Trying to update key that is not present in formData.');
    }
  };

  getThisFormDataValues = () => {
    const formDataValues = {};
    Object.keys(this.state.formData).forEach(key => { formDataValues[key] = this.state.formData[key].value; });
    return formDataValues;
  };

  formData = () => ({
    ...this.props.navigation.getParam('formData', {}),
    ...this.getThisFormDataValues(),
  });

  isValid = () => {
    return Object.values(this.state.formData).reduce((valid, currentData) => valid && currentData.valid, true);
  };

  submit = () => {
    if (this.isValid()) {
      this.setState({ containErrors: false });
      this.props.navigation.navigate('SendReport', { formData: this.formData() })
    } else {
      this.setState({ containErrors: true });
    }
  };

  render() {
    const validationMessage = (
      <Text style={{ color: RED }}>
        Sentimos falta de algumas informações obrigatórias. Por favor, preencha-as e tente novamente.
      </Text>
    );

    return (
      <View>
        <Picker
          required
          testID="gender-input"
          placeholder="Gênero"
          showValidation={this.state.containErrors}
          value={this.state.formData.gender.value}
          onValueChange={(value, valid) => this.updateFormDataValue('gender', value, valid)}
          items={this.props.data.formOptions.gender_options}
        />
        <Picker
          required
          testID="skin-color-input"
          placeholder="Cor"
          showValidation={this.state.containErrors}
          value={this.state.formData.skinColor.value}
          onValueChange={(value, valid) => this.updateFormDataValue('skinColor', value, valid)}
          items={this.props.data.formOptions.skin_color_options}
        />
        <Picker
          testID="age-input"
          placeholder="Idade"
          showValidation={this.state.containErrors}
          value={this.state.formData.age.value}
          onValueChange={(value, valid) => this.updateFormDataValue('age', value, valid)}
          items={this.props.data.formOptions.age_options}
        />
        <Picker
          testID="orientation-input"
          placeholder="Orientação Sexual"
          showValidation={this.state.containErrors}
          value={this.state.formData.sexualOrientation.value}
          onValueChange={(value, valid) => this.updateFormDataValue('sexualOrientation', value, valid)}
          items={this.props.data.formOptions.sexual_orientation}
        />
        <Picker
          testID="wage-input"
          placeholder="Renda Aproximada"
          showValidation={this.state.containErrors}
          value={this.state.formData.wage.value}
          onValueChange={(value, valid) => this.updateFormDataValue('wage', value, valid)}
          items={this.props.data.formOptions.wage_options}
        />
        <SectionHeader title="Saiba mais" />
        <SectionText title="Acompanhe relatos da mesma empresa ou cidade." />
        <TextInput
          required
          testID="email-input"
          placeholder="Email"
          textContentType="emailAddress"
          keyboardType="email-address"
          showValidation={this.state.containErrors}
          onValueChange={(email, valid) => this.updateFormDataValue('email', email, valid)}
          value={this.state.formData.email.value}
        />
        <TextInput
          testID="name-input"
          placeholder="Nome"
          showValidation={this.state.containErrors}
          onValueChange={(name, valid) => this.updateFormDataValue('name', name, valid)}
          value={this.state.formData.name.value}
        />
        <CheckBox
          onValueChange={() => this.setState({ acceptedPolicies: !this.state.acceptedPolicies })}
          value={this.state.acceptedPolicies}
          label="Ao registrar esse relato eu concordo com a Política de Privacidade do Eu Também."
        />
        <Text style={appStyles.link} onPress={() => this.props.navigation.navigate('PrivacyPolicy')}>
          Acessar Política de Privacidade
        </Text>
        { this.state.containErrors ? validationMessage : null }
        <Button
          testID="send-button"
          onPress={() => this.submit()}
          title="Prosseguir"
        />
      </View>
    );
  }
}

const valueLabelShape = PropTypes.arrayOf(PropTypes.shape({
  value: PropTypes.string,
  label: PropTypes.string,
})).isRequired;

AboutYouForm.propTypes = {
  navigation: PropTypes.shape({
    navigate: PropTypes.func,
    getParam: PropTypes.func,
  }).isRequired,
  data: PropTypes.shape({
    formOptions: PropTypes.shape({
      skin_color_options: valueLabelShape,
      age_options: valueLabelShape,
      wage_options: valueLabelShape,
      gender_options: valueLabelShape,
      followup_actions_options: valueLabelShape,
      yes_no_optional_options: valueLabelShape,
      sexual_orientation: valueLabelShape,
    }).isRequired,
  }).isRequired,
};

export default formWrapper(loading(AboutYouForm));
