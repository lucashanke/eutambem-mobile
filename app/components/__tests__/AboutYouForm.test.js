import 'react-native';
import React from 'react';
import { shallow } from 'enzyme';
import sinon from 'sinon';

import { AboutYouForm } from '../AboutYouForm';

export const GENDER_OPTIONS = [{
  value: 'woman',
  label: 'Mulher',
}, {
  value: 'man',
  label: 'Homem',
}, {
  value: 'nonbinary',
  label: 'Não binário',
}];

export const SKIN_COLOR_OPTIONS = [{
  value: 'black',
  label: 'Negra',
}, {
  value: 'white',
  label: 'Branca',
}, {
  value: 'brown',
  label: 'Parda',
}, {
  value: 'asian',
  label: 'Amarela',
}, {
  value: 'indian',
  label: 'Indígena',
}, {
  value: 'notInformed',
  label: 'Prefiro não dizer',
}];

export const AGE_OPTIONS = [{
  value: 'lessThan18',
  label: 'Menor de 18 anos',
}, {
  value: 'between18and25',
  label: 'Entre 18 e 25 anos',
}, {
  value: 'between26and35',
  label: 'Entre 26 e 35 anos',
}, {
  value: 'between36and50',
  label: 'Entre 36 e 50 anos',
}, {
  value: 'between51and65',
  label: 'Entre 51 e 65 anos',
}, {
  value: 'over65',
  label: 'Maior de 65 anos',
}];

export const WAGE_OPTIONS = [{
  value: 'belowOne',
  label: 'Até R$724,00',
}, {
  value: 'fromOneToThree',
  label: 'De R$724,01 a R$2.172,00',
}, {
  value: 'fromThreeToFive',
  label: 'De R$2.172,01 a R$3.620,00',
}, {
  value: 'fromFiveToTen',
  label: 'De R$3.620,01 a R$7.240,00',
}, {
  value: 'fromTenToTwenty',
  label: 'De R$7.240,01 a R$14.480,00',
}, {
  value: 'overTwenty',
  label: 'Mais de R$14.480,01',
}];

const props = {
  navigation: {
    navigate: sinon.spy(),
  },
  data: {
    formOptions: {
      wage_options: WAGE_OPTIONS,
      age_options: AGE_OPTIONS,
      skin_color_options: SKIN_COLOR_OPTIONS,
      gender_options: GENDER_OPTIONS,
    },
  },
};

describe('AboutYouForm', () => {
  it('renders correctly', () => {
    const wrapper = shallow(<AboutYouForm {...props} />);
    expect(wrapper).toMatchSnapshot();
  });

  it('passes the form data to the SendReport screen', () => {
    props.navigation.getParam = (key) => {
      if (key === 'formData') {
        return {
          description: 'something',
        };
      }
      return null;
    };

    const wrapper = shallow(<AboutYouForm {...props} />);
    const initialFormData = wrapper.state().formData;
    const initialFormDataValues = wrapper.instance().getThisFormDataValues();
    wrapper.setState({
      formData: {
        ...initialFormData,
        email: { value: 'someone@example.com' },
      },
    });

    wrapper.instance().isValid = () => true;

    wrapper.find('Button').simulate('press');

    expect(props.navigation.navigate.calledWith('SendReport', {
      formData: {
        description: 'something',
        ...initialFormDataValues,
        email: 'someone@example.com',
      },
    })).toBeTruthy();
  });
});
