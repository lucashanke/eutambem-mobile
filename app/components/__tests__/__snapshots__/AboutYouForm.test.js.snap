// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AboutYouForm renders correctly 1`] = `
<ForwardRef(View)>
  <Input
    items={
      Array [
        Object {
          "label": "Mulher",
          "value": "woman",
        },
        Object {
          "label": "Homem",
          "value": "man",
        },
        Object {
          "label": "Não binário",
          "value": "nonbinary",
        },
      ]
    }
    onValueChange={[Function]}
    placeholder="Gênero"
    required={true}
    showValidation={false}
    value=""
  />
  <Input
    items={
      Array [
        Object {
          "label": "Negra",
          "value": "black",
        },
        Object {
          "label": "Branca",
          "value": "white",
        },
        Object {
          "label": "Parda",
          "value": "brown",
        },
        Object {
          "label": "Amarela",
          "value": "asian",
        },
        Object {
          "label": "Indígena",
          "value": "indian",
        },
        Object {
          "label": "Prefiro não dizer",
          "value": "notInformed",
        },
      ]
    }
    onValueChange={[Function]}
    placeholder="Cor"
    required={true}
    showValidation={false}
    value=""
  />
  <Input
    items={
      Array [
        Object {
          "label": "Menor de 18 anos",
          "value": "lessThan18",
        },
        Object {
          "label": "Entre 18 e 25 anos",
          "value": "between18and25",
        },
        Object {
          "label": "Entre 26 e 35 anos",
          "value": "between26and35",
        },
        Object {
          "label": "Entre 36 e 50 anos",
          "value": "between36and50",
        },
        Object {
          "label": "Entre 51 e 65 anos",
          "value": "between51and65",
        },
        Object {
          "label": "Maior de 65 anos",
          "value": "over65",
        },
      ]
    }
    onValueChange={[Function]}
    placeholder="Idade"
    required={false}
    showValidation={false}
    value=""
  />
  <Input
    items={
      Array [
        Object {
          "label": "Até R$724,00",
          "value": "belowOne",
        },
        Object {
          "label": "De R$724,01 a R$2.172,00",
          "value": "fromOneToThree",
        },
        Object {
          "label": "De R$2.172,01 a R$3.620,00",
          "value": "fromThreeToFive",
        },
        Object {
          "label": "De R$3.620,01 a R$7.240,00",
          "value": "fromFiveToTen",
        },
        Object {
          "label": "De R$7.240,01 a R$14.480,00",
          "value": "fromTenToTwenty",
        },
        Object {
          "label": "Mais de R$14.480,01",
          "value": "overTwenty",
        },
      ]
    }
    onValueChange={[Function]}
    placeholder="Renda Aproximada"
    required={false}
    showValidation={false}
    value=""
  />
  <SectionHeader
    title="Saiba mais"
  />
  <SectionText
    title="Acompanhe relatos da mesma empresa ou cidade."
  />
  <Input
    keyboardType="email-address"
    onValueChange={[Function]}
    placeholder="Email"
    required={true}
    showValidation={false}
    textContentType="emailAddress"
    value=""
  />
  <Input
    onValueChange={[Function]}
    placeholder="Nome"
    required={false}
    showValidation={false}
    value=""
  />
  <Input
    label="Ao registrar esse relato eu concordo com a Política de Privacidade do Eu Também."
    onValueChange={[Function]}
    required={false}
    value={false}
  />
  <ForwardRef(Text)
    onPress={[Function]}
    style={
      Object {
        "alignSelf": "flex-end",
        "color": "#91508c",
        "marginBottom": 10,
        "marginTop": 10,
      }
    }
  >
    Acessar Política de Privacidade
  </ForwardRef(Text)>
  <Button
    onPress={[Function]}
    title="Enviar Relato"
    type="confirm"
  />
</ForwardRef(View)>
`;
